### PR TITLE
Corey/testing

### PR DESCRIPTION
**Pull request by PoLoMoTo**
Features added:
- Automated testing for all views
- Integrate testing into GitHub action, run on every push
- Separate launch modes for development and production
- Database construction and tear down for testing

Workflow changes:
- To run the server in development mode use `npm run dev`
- Run the testing suites with `npm test`
- Github will not allow merges to master when tests have failed

Work remaining:
- Add tests for form submissions
- Database tear down and rebuild for each test suite

Creating tests:
- All test suites are in the 'test' directory, one for every general aspect of the project
- See the sample test for an example of how testing is done for Jest as well as [Jest's documentation](https://jestjs.io/docs/en/getting-started)
- See the existing tests for views for how to use Supertest to test routes as well as [Supertest's documentation](https://github.com/visionmedia/supertest#readme)
- Please add tests for all new routes and modify existing tests if you change a route

### CODE REVIEW

File: database.js [hunk 1]
```diff
@@ -927,10 +922,6 @@ async function queryRoleByID(id){
   return role;
 }
 
-async function queryAllRoles(){
```
Comment (JordanH821) [inline database.js:930]: Should not be deleted

File: routes/settings-routes.js [hunk 2]
```diff
@@ -5,12 +5,13 @@ const utils = require('../utils');
 const rbac = require('../rbac');
 logger.module = 'settings-routes';
 
-
 const authCheck = (req, res, next) => {
-  if(!req.user){
+  if(!req.user && process.env.TESTING == undefined){
       // if not logged in
+      console.log('bouncing');
```
Comment (JordanH821) [inline routes/settings-routes.js:11]: change to logger, unless this is before the logger module

File: server.js [hunk 3]
```diff
@@ -26,60 +37,68 @@ app.use(bodyParser.urlencoded({
...
 
 // setup routes
 app.use('/auth', authRoutes);
 app.use('/settings', settingsRoutes);
 
 const authCheck = (req, res, next) => {
-  if(!req.user){
+  if(!req.user && process.env.TESTING == undefined){
       // if not logged in
+      console.log('bouncing');
```
Comment (JordanH821) [inline server.js:59]: logger

File: server.js [hunk 4]
```diff
@@ -26,60 +37,68 @@ app.use(bodyParser.urlencoded({
...
       res.redirect('/auth/google');
   } else {
+      console.log('success');
```
Comment (JordanH821) [inline server.js:62]: logger

**Comment by davidem00** [comment]
in the list of tests above, do the Green Checks, Red 'X's and no marks refer to a test being enabled or not, or to whether it passed or failed for the given pull request merge?

File: database.js (hunk 1 above)
Comment (PoLoMoTo) [reply database.js:930]: It is a duplicate function.

**Comment by PoLoMoTo** [comment]
> A few comments about logger and a deleted db method, other than that looks good

Please see new changes

> in the list of tests above, do the Green Checks, Red 'X's and no marks refer to a test being enabled or not, or to whether it passed or failed for the given pull request merge?

I believe you are looking at the list of commits.  The check mark and X on commits is whether all tests passed or if some failed respectively.  Currently running tests is represented with an orange/yellow circle.  Tests are only run on the latest commit when pushed to Github, any that do not specifically get tested do not get a check or an 'X'.

File: routes/settings-routes.js (hunk 2 above)
Comment (PoLoMoTo) [reply routes/settings-routes.js]: I just removed them as they're no longer in master.

**Comment by davidem00** [comment]
So the reason for a commit being skipped for testing (no mark shown) is that it is being merged as part of a chain of other commits for which a later one will run the same suite of tests?

**Comment by PoLoMoTo** [comment]
Correct, more so pushed than merged though.

**Pull request merged**